File: Data_Structure/UnionFind.cpp
// verified with:
// https://judge.yosupo.jp/submission/103478

#include<bits/stdc++.h>
using namespace std;

// αはアッカーマンの逆関数
struct UnionFind{
  UnionFind() : n(0) {}
  explicit UnionFind(int n) : n(n),parents(n,-1) {}
  
  // 頂点xと頂点yをmerge <ならしO(α(n))>
  void merge(int x, int y){
    x = root(x), y = root(y);
    if (x == y) return;
    if (parents[x] > parents[y]) swap(x,y);
    parents[x] += parents[y];
    parents[y] = x;
  }

  // 頂点xと頂点yが連結か否かを判定 <ならしO(α(n))>
  bool same(int x, int y){
    return root(x) == root(y);
  }

  // 頂点xの親を返す <ならしO(α(n))>
  int root(int x){
    if (parents[x] < 0){
      return x;
    }else{
      return parents[x] = root(parents[x]);
    }
  }

  // 頂点xが属する連結成分の大きさを返す <ならしO(α(n))>
  int size(int x){
    return -parents[root(x)];
  }
  
  // 連結成分ごとにまとめて返す <ならしO(n)>
  vector<vector<int>> groups(){
    vector<vector<int>> ret(n);
    vector<int> cnt(n),roots(n);
    for(int i = 0; i < n; ++i) {
      roots[i] = root(i);
      ++cnt[roots[i]];
    }
    for(int i = 0; i < n; ++i) ret[i].reserve(cnt[i]);
    for(int i = 0; i < n; ++i) ret[roots[i]].push_back(i);
    ret.erase(remove_if(ret.begin(), ret.end(), [&](const std::vector<int>& i) {return i.empty();}), ret.end());
    return ret;
  }

  private:
    // 頂点の個数
    int n;
    // 負なら連結成分の大きさ、非負なら親
    vector<int> parents;  
};

File: Data_Structure/UnionFind.py
# verified with:
# https://judge.yosupo.jp/submission/103479

# αはアッカーマンの逆関数
class UnionFind:
    # コンストラクタ
    def __init__(self, n):
        self.n = n   # 頂点の個数
        self.parents = [-1]*n   # 負なら連結成分の大きさ、非負なら親
    
    # 頂点xと頂点yをmerge <ならしO(α(n))>
    def merge(self,x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return
        if self.parents[x] > self.parents[y]:
            x,y = y,x
        self.parents[x] += self.parents[y]
        self.parents[y] = x
        
    # 頂点xと頂点yが連結か否かを判定 <ならしO(α(n))>
    def same(self,x, y):
        x = self.root(x)
        y = self.root(y)
        return x == y

    # 頂点xの親を返す <ならしO(α(n))>
    def root(self,x):
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.root(self.parents[x])
            return self.parents[x]
    
    # 頂点xが属する連結成分の大きさを返す <ならしO(α(n))>
    def size(self,x):
        return -self.parents[self.root(x)]
        
    # 連結成分ごとにまとめて返す <O(n)>
    def groups(self):
        ret = [[] for i in range(self.n)]
        for i in range(self.n):
            ret[self.root(i)].append(i)
        return [i for i in ret if len(i)]